src/sdk/namespace_fs.js

```diff
@@ -1560,6 +1625,78 @@ class NamespaceFS {
     }
+    //////////////////////////
+    //// VERSIONING UTILS ////
+    //////////////////////////
+
+    _is_versioning_enabled() {
+        return this.versioning === ver_enabled;
+    }
+
+    _get_version_id_by_stat({ino, mtimeSec, mtimeNsec}) {
+        // TODO: GPFS might require generation number to be added to version_id
+       return 'mtime-' + mtimeSec.toString() + '-' + mtimeNsec.toString() + '-ino-' + ino.toString();
+    }
+
+    // 1. if version_id_str is null version - nothing to etract
+    // 2. else extract the attributes from it or fail for invalid version_id_str
+    // version_id_str - mtime-{mtimeSec}-{mtimeNsec}-ino-{ino} | explicit null
+    // returns mtimeSec, mtimeNsec, ino (inode_number)
+    _extract_version_attr_from_xattr(version_id_str) {
+        if (version_id_str === 'null') return;
+        const arr = (version_id_str.match(num_regexp) || []).map(n => Number(n));
+        if (arr.length < 3) throw new Error('Invalid version_id_string, cannot extact attributes');
+        return { mtimeSec: arr[0], mtimeNsec: arr[1], ino: arr[2] };
+    }
+
+    _get_version_id_by_xattr(stat) {
+       return (stat && stat.xattr[XATTR_VERSION_ID]) || 'null';
+    }
+
+    _get_version_path(key, version_id) {
+        const key_version = path.basename(key) + (version_id ? '_' + version_id : '');
+        return path.normalize(path.join(this.bucket_path, path.dirname(key), hidden_versions_path, key_version));
+    }
+
+    // this function returns the following version attributes -
+    // version_id_str - mtime-{mtimeSec}-{mtimeNsec}-ino-{ino} | explicit null
+    // mtimeSec - modified timestmap seconds - last time the content of the file was modified
+    // mtimeNsec - modified timestmap nano seconds - last time the content of the file was modified
+    // ino - refers to the data stored in a particular location
+    // if xattr contains version attr - return attr of xattr
+    // else - it's a null version - return stat
+    async _get_version_attr(fs_context, version_path) {
+        try {
+            const stat = await nb_native().fs.stat(fs_context, version_path);
+            dbg.log1('NamespaceFS._get_version_attr stat ', stat);
+
+            const version_id_str = this._get_version_id_by_xattr(stat);
+            const ver_attr_by_xattr = this._extract_version_attr_from_xattr(version_id_str);
+            return { ...(ver_attr_by_xattr || stat), version_id_str };
+        } catch (err) {
+            if (err.code !== 'ENOENT') throw err;
+            dbg.error('NamespaceFS._get_version_attr version doesn\'t exist', err);
+        }
+        // if stat failed, undefined will return
+    }
+
+    // 1. if version exists in .versions/ folder - return its path
+    // 2. else if version is latest version - return latest version path 
+    // 3. throw ENOENT error
+    async _find_version_path(fs_context, { key, version_id }) {
+        const cur_ver_path = this._get_file_path({ key });
+        if (version_id) {
```

Comment:
style: I would check if not version_id and return early so the main flow keeps flat in the function
```
if (!version_id) return cur_ver_path;
...
```